server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -1175,43 +1153,15 @@ protected E lockAndLoadById(Class<E> entityClass, Serializable id) {
                 }
             }
 
-            // Build a query to fetch the remaining entities
+            // Fetch the remaining entities from the DB
             if (idSet.size() > 0) {
-                // Get the entity's metadata so we can ask Hibernate for the name of its identifier
-                String idName = metadata.getIdentifierPropertyName();
-                if (idName == null) {
-                    // This shouldn't happen.
-                    throw new RuntimeException("Unable to fetch identifier property name");
-                }
-
-                // Impl note:
-                // We're building the query here using JPA Criteria to avoid fiddling with string
-                // building and, potentially, erroneously using the class name as the entity name.
-                // Additionally, using a query (as opposed to the .find and .load methods) lets us set
-                // the flush, cache and lock modes for the entity we're attempting to fetch.
-                CriteriaBuilder builder = entityManager.getCriteriaBuilder();
-                CriteriaQuery<E> query = builder.createQuery(entityClass);
-                Root<E> root = query.from(entityClass);
-                Path<Serializable> target = root.<Serializable>get(idName);
-                ParameterExpression<List> param = builder.parameter(List.class);
-
-                query.select(root).where(target.in(param));
-
-                // Note that it's critical here to set both modes, as Hibernate is wildly inconsistent
-                // (and non-standard) in which properties it actually accepts when processing its own
-                // config objects. The cache mode combination specified below ends up being evaluated
-                // by Hibernate down to a CacheMode.REFRESH.
-                TypedQuery<E> executable = entityManager.createQuery(query)
-                    .setFlushMode(FlushModeType.COMMIT)
-                    .setHint(AvailableSettings.SHARED_CACHE_RETRIEVE_MODE, CacheRetrieveMode.BYPASS)
-                    .setHint(AvailableSettings.SHARED_CACHE_STORE_MODE, CacheStoreMode.REFRESH)
-                    .setLockMode(LockModeType.PESSIMISTIC_WRITE);
-
-                // Step through the query in blocks
-                for (List<Serializable> block : Iterables.partition(idSet, getBatchBlockSize())) {
-                    executable.setParameter(param, block);
-                    result.addAll(executable.getResultList());
-                }
+                result.addAll(((Session) session)
+                    .byMultipleIds(this.entityType())
+                    .with(new LockOptions(LockMode.PESSIMISTIC_WRITE))
+                    .with(CacheMode.REFRESH)
+                    .enableSessionCheck(false)
+                    .enableOrderedReturn(false)
+                    .multiLoad(new ArrayList(idSet)));
```

Comment:
With this new query, we no longer have to partition the ids and make multiple lookups?

Reply:
Nope, it handles that for us internally.

Reply:
Beauty! :+1: